### DISCUSSION TITLE
Smart pointers

### DISCUSSION THREAD

**Discussion by domcir** [bytedeco/javacpp#623] [Q&A] [answered]
Hi

The spart pointer adapter works only for std::shared_ptr<MyObject> but not for std::shared_ptr<const MyObject>.
Is there a way to solve my issue? Can I just ease a modified adapter where I add const here and there?

Also how can I map an internal type in the library created from a macro below to a smart pointer type.

#define CTP_DECLARE_PTR(x)\
	using x##Ptr	= std::shared_ptr<x>;\
	using x##CPtr	= std::shared_ptr<x const>;\

where "MyObjectCPtr" type in C++ is actually "std::shared_ptr<const MyObject>" but it's not picked up by the parser and a type MyObjectCPtr is used instead.

Thanks
Dom

**Comment by saudet** [comment, marked as answer]
We can still use `@SharedPtr` on those types, it will work fine.

**Reply by domcir** [reply]
It doesn't work, the C++ compilation fails as [here ](https://github.com/bytedeco/javacpp/blob/master/src/main/java/org/bytedeco/javacpp/tools/Generator.java#L1349) the const keyword is missing.

**Reply by saudet** [reply]
We can also use `@Cast` as appropriate to match these keywords.

**Reply by domcir** [reply]
Do you  mean to define manually the signature and add @Cast where required?
Also how could I map the macro? I'm not sure if I could use [that](https://github.com/bytedeco/javacpp/wiki/Mapping-Recipes#specifying-names-to-use-in-java).

Thanks in advance
Dom

**Reply by saudet** [reply]
Yes, something like that. There are a few examples close to yours here I believe:
https://github.com/bytedeco/javacpp-presets/blob/master/pytorch/src/main/java/org/bytedeco/pytorch/presets/torch.java
We don't need to map the macros, it should be fine if you just put them after std::shared_ptr<...> as "cppTypes" like `new Info("std::shared_ptr<MyObject>", "MyObjectPtr", "MyObjectCPtr")...`.

**Reply by domcir** [reply]
Thanks, I'll try it and let  you know.

**Reply by saudet** [reply]
Does it work?

**Reply by domcir** [reply]
yes, thanks!
